src/masternode/activemasternode.cpp

```diff
@@ -201,10 +202,11 @@ bool CActiveMasternodeManager::GetLocalAddress(CService& addrRet)
     if (!fFoundLocal) {
         bool empty = true;
         // If we have some peers, let's try to find our local address from one of them
-        g_connman->ForEachNodeContinueIf(CConnman::AllNodes, [&fFoundLocal, &empty](CNode* pnode) {
+        auto service = WITH_LOCK(activeMasternodeInfoCs, return activeMasternodeInfo.service);
+        g_connman->ForEachNodeContinueIf(CConnman::AllNodes, [&](CNode* pnode) {
             empty = false;
             if (pnode->addr.IsIPv4())
-                fFoundLocal = GetLocal(activeMasternodeInfo.service, &pnode->addr) && IsValidNetAddr(activeMasternodeInfo.service);
+                fFoundLocal = GetLocal(activeMasternodeInfo.service, &pnode->addr) && IsValidNetAddr(service);
```

Comment:
```suggestion
                fFoundLocal = GetLocal(service, &pnode->addr) && IsValidNetAddr(service);
```

Reply:
Applied this and also your previous changes:

- done d876781, excluding changes where we pass temporary copies by values and do possibly do "old "data reads with no benefit. I kept only the changes that could trigger dead locks on the network and moved locks out of for loops that are quite big.  
- done 099909a
- done da44228